AdyenActions/Components/3DS2/Action handlers/3DS2 Without Delegated Authentication/ThreeDS2CoreActionHandler.swift

```diff
@@ -140,34 +156,68 @@ internal class ThreeDS2CoreActionHandler: AnyThreeDS2CoreActionHandler {
                                                          threeDSRequestorAppURL: threeDSRequestorAppURL ?? token.threeDSRequestorAppURL)
         transaction.performChallenge(with: challengeParameters) { [weak self] challengeResult, error in
             guard let result = challengeResult else {
-                let error = error ?? UnknownError(errorDescription: "Both error and result are nil, this should never happen.")
-                self?.didFail(with: error, completionHandler: completionHandler)
+                self?.didReceiveErrorOnChallenge(error: error, challengeAction: challengeAction, completionHandler: completionHandler)
                 return
             }
 
             self?.didFinish(with: result,
                             authorizationToken: challengeAction.authorisationToken,
                             completionHandler: completionHandler)
         }
-
+    }
+    
+    /// Invoked to handle the error flow of a challenge handling by the 3ds2sdk.
+    /// For challenge cancelled we return the control back to the merchant immediately as an error.
+    private func didReceiveErrorOnChallenge(error: Error?,
+                                            challengeAction: ThreeDS2ChallengeAction,
+                                            completionHandler: @escaping (Result<ThreeDSResult, Error>) -> Void) {
+        guard let error = error as? NSError else {
+            didFail(with: UnknownError(errorDescription: "Both error and result are nil, this should never happen."),
+                    completionHandler: completionHandler)
+            return
+        }
+        switch (error.domain, error.code) {
+        case (ADYRuntimeErrorDomain, Int(ADYRuntimeErrorCode.challengeCancelled.rawValue)):
+            didFail(with: error,
+                    completionHandler: completionHandler)
+        default:
+            didFinish(threeDS2SDKError: error.base64Representation(),
+                      authorizationToken: challengeAction.authorisationToken,
+                      completionHandler: completionHandler)
+        }
+    }
+    
+    private func didFinish(threeDS2SDKError: String,
+                           authorizationToken: String?,
+                           completionHandler: @escaping (Result<ThreeDSResult, Error>) -> Void) {
+        do {
+            // When we get an error we need to send transStatus as "U" along with the threeDS2SDKError field.
+            let threeDSResult = try ThreeDSResult(authorizationToken: authorizationToken,
```

Comment:
A weird one eh :) Let's create a private constant enum in this class and use that instead.

Reply:
Yeah :) a limitation in the backend to process the errors.  Done - with the Constant enum.